### PR TITLE
Use a single codegen unit for fully optimized builds.

### PR DESCRIPTION
**Pull request by nnethercote**
r? @ghost

### CODE REVIEW

**Comment by nnethercote** [comment]
This gives fantastic results for me locally, with walltime improvements up to 6%, and max-rss improvements of up to 8%.

@bors try @rust-timer queue

File: src/bootstrap/config.rs [hunk 1]
```diff
@@ -1368,6 +1370,16 @@ impl Config {
         config.llvm_plugins = llvm_plugins.unwrap_or(false);
         config.rust_optimize = optimize.unwrap_or(true);
 
+        let fully_optimized = config.rust_optimize && debug != Some(true) && !config.incremental;
+        if fully_optimized && rust_codegen_units.is_none() {
+            rust_codegen_units = Some(1)
+        }
+        if fully_optimized && rust_codegen_units_std.is_none() {
+            rust_codegen_units_std = Some(1)
```
Comment (jyn514) [inline src/bootstrap/config.rs:1378]: this is a pre-existing bug and doesn't affect your pr, but i think this should default to `rust_codegen_units` if it's set, not have an independent default, like how `debug-assertions-std` works

**Review by jyn514 — COMMENTED** [review]
this is really exciting!

**Comment by lqd** [comment]
We had been experimenting with this in https://github.com/rust-lang/rust/pull/107651, where we've encountered the same CI issue as this PR (likely BOLT's https://github.com/llvm/llvm-project/issues/59174). We'll keep working on this, and circumventing the issue in that other PR.

**Comment by nnethercote** [comment]
Yes, let's continue this work in #107651.

**Pull request closed without merging**